pkg/image/importer/importer.go

```diff
@@ -0,0 +1,739 @@
+package importer
+
+import (
+	"encoding/json"
+	"fmt"
+	"net/http"
+	"net/url"
+	"path"
+	"strings"
+	"time"
+
+	"github.com/golang/glog"
+	gocontext "golang.org/x/net/context"
+
+	"github.com/docker/distribution"
+	"github.com/docker/distribution/context"
+	"github.com/docker/distribution/digest"
+	"github.com/docker/distribution/manifest/schema1"
+	"github.com/docker/distribution/registry/api/errcode"
+	"github.com/docker/distribution/registry/api/v2"
+	registryclient "github.com/docker/distribution/registry/client"
+	"github.com/docker/distribution/registry/client/auth"
+	"github.com/docker/distribution/registry/client/transport"
+
+	kapi "k8s.io/kubernetes/pkg/api"
+	kapierrors "k8s.io/kubernetes/pkg/api/errors"
+	"k8s.io/kubernetes/pkg/api/unversioned"
+	client "k8s.io/kubernetes/pkg/client/unversioned"
+	"k8s.io/kubernetes/pkg/util"
+	"k8s.io/kubernetes/pkg/util/fielderrors"
+	"k8s.io/kubernetes/pkg/util/sets"
+
+	"github.com/openshift/origin/pkg/dockerregistry"
+	"github.com/openshift/origin/pkg/image/api"
+	"github.com/openshift/origin/pkg/image/api/dockerpre012"
+)
+
+const (
+	// Add a dockerregistry.Client to the passed context with this key to support v1 Docker registry importing
+	ContextKeyV1RegistryClient = "v1-registry-client"
+)
+
+// Interface loads images into an image stream import request.
+type Interface interface {
+	Import(cxt *AuthContext, isi *api.ImageStreamImport) error
+}
+
+// ImageStreamImport implements an import strategy for Docker images. It keeps a cache of images
+// per distinct auth context to reduce duplicate loads
+type ImageStreamImport struct {
+	maximumTagsPerRepo int
+
+	limiter util.RateLimiter
+
+	imageCache map[*AuthContext]map[location]*api.Image
+}
+
+// NewImageStreamImport creates an importer that will load images from a remote Docker registry into an
+// ImageStreamImport object. Limiter may be nil.
+func NewImageStreamImport(maximumTagsPerRepo int, limiter util.RateLimiter) *ImageStreamImport {
+	if limiter == nil {
+		limiter = util.NewFakeRateLimiter()
+	}
+	return &ImageStreamImport{
+		maximumTagsPerRepo: maximumTagsPerRepo,
+
+		limiter: limiter,
+
+		imageCache: make(map[*AuthContext]map[location]*api.Image),
+	}
+}
+
+// Import tries to complete the provided isi object with images loaded from remote registries.
+func (i *ImageStreamImport) Import(ctx *AuthContext, isi *api.ImageStreamImport) error {
+	cache := i.imageCache[ctx]
+	if cache == nil {
+		cache = make(map[location]*api.Image)
+		i.imageCache[ctx] = cache
+	}
+	importImages(ctx, isi, cache, i.limiter)
+	importFromRepository(ctx, isi, i.maximumTagsPerRepo, cache, i.limiter)
+	return nil
+}
+
+func importImages(ctx *AuthContext, isi *api.ImageStreamImport, cache map[location]*api.Image, limiter util.RateLimiter) {
```

Comment:
> Seems like *imageStreamImport ought to be a method receiver.

Oh sweet jesus.  Its the same type name that's a completely different thing in a different object.  Passing an `ImageStreamImport` to an `ImageStreamImport` is insane.  `ImageStreamImporter` seems more reasonable.